File: ruby-version/lib/astromapper/rules/ruleset.rb
require 'yaml'
require_relative 'expr'

module Astromapper
  module Rules
    # A Ruleset is the data-driven definition of a generation system (Traveller 5,
    # Cepheus, house rules…), loaded from `rules/<name>.yml`. For now it owns the
    # trade-classification table; UWP steps, starport/base tables, and the algorithmic
    # module hooks (extensions/climate/native) follow in later phases.
    class Ruleset
      # Built-in rulesets ship in ruby-version/rules; projects may add their own.
      BUILTIN_DIR = File.expand_path('../../../rules', __dir__)

      attr_reader :name, :data

      # Human-facing name (the YAML `name:` field), falling back to the file slug.
      def title
        @data['name'] || @name
      end

      def self.search_dirs(root = nil)
        dirs = [BUILTIN_DIR]
        dirs.unshift(File.join(root, 'rules')) if root
        dirs
      end

      def self.load(name, root: nil)
        file = search_dirs(root).map { |d| File.join(d, "#{name}.yml") }.find { |f| File.file?(f) }
        raise "Unknown ruleset #{name.inspect} (looked in #{search_dirs(root).join(', ')})" unless file
        data = YAML.safe_load(File.read(file)) || {}
        if (base = data['extends'])               # shallow inheritance: child keys win
          data = deep_merge(load(base, root: root).data, data)
        end
        new(name, data).validate!
      end

      # Child keys win. A child key with a trailing "!" replaces the parent value
      # wholesale instead of deep-merging — needed when a ruleset has a *different*
      # set (e.g. Cepheus trade codes), not additions to the parent's.
      def self.deep_merge(a, b)
        result = a.dup
        b.each do |k, bv|
          if k.to_s.end_with?('!')
            result[k.to_s.chomp('!')] = bv
          elsif result[k].is_a?(Hash) && bv.is_a?(Hash)
            result[k] = deep_merge(result[k], bv)
          else
            result[k] = bv
          end
        end
        result
      end

      def initialize(name, data)
        @name = name
        @data = data
        @trade = (data['trade_codes'] || {}).transform_values { |cond| Expr.compile(cond) }
      end

      # Fail fast on a malformed ruleset (typo'd custom file, broken inheritance), with
      # a message that names every problem at once. Returns self so load can chain it.
      def validate!
        errors = []
        errors << "missing `hex` alphabet" unless @data['hex'].is_a?(String) && !@data['hex'].empty?
        %w[size atmo hydro pop gov law].each do |step|
          s = (@data['uwp'] || {})[step]
          errors << "uwp.#{step}: missing or has no `roll`" unless s.is_a?(Hash) && s['roll']
        end
        errors << "missing `starport.table`" unless @data.dig('starport', 'table').is_a?(Array)
        %w[extensions climate native].each { |slot| module_for(slot) }   # raises on a bad name
        raise "ruleset #{@name.inspect} is invalid:\n- #{errors.join("\n- ")}" unless errors.empty?
        self
      end

      # Trade classifications for a world. `ctx` is a Hash of UWP values keyed by name
      # (size, atmo, hydro, pop, gov, law, tech, port, temp). Order follows the YAML.
      def trade_codes(ctx)
        @trade.select { |_code, test| test.call(ctx) == true }.keys
      end

      # Starport letter for an orientation roll (clamped to the table). Deterministic.
      def starport(roll)
        table = @data.dig('starport', 'table') || []
        table[roll.to_i.clamp(0, table.size - 1)]
      end

      # Summed Tech-Level DM from the port map + per-digit arrays. Deterministic.
      def tech_dm(ctx)
        t = @data['tech_dm'] || {}
        dm = (t['port'] || {})[ctx['port']].to_i
        %w[size atmo hydro pop gov].each { |k| dm += ((t[k] || [])[ctx[k].to_i] || 0) }
        dm
      end

      # 2D threshold for a base at this starport, or nil if that port can't have it.
      def base_threshold(kind, port)
        (@data.dig('bases', kind) || {})[port]
      end

      # Does a base roll qualify? T5 wants `roll <= threshold` (default); Cepheus and
      # classic Traveller want `>=`. Set `bases.op` in the ruleset to switch.
      def base_meets?(roll, threshold)
        Expr.compare(@data.dig('bases', 'op') || '<=', roll, threshold)
      end

      # Name of the code module wired to an algorithmic slot (extensions/climate/native).
      # Returns a bare name like "t5" or "none"; defaults to "t5" when unspecified. The
      # name is constrained to a word so it can only resolve to a `<slot>_<name>` method.
      def module_for(slot)
        name = (@data['modules'] || {})[slot].to_s.downcase
        name = 't5' if name.empty?
        raise "ruleset #{@name.inspect}: bad module name #{name.inspect} for #{slot}" unless name =~ /\A\w+\z/
        name
      end

      # Evaluate one UWP step (size, atmo, hydro, pop, gov, law) against `ctx` — a Hash
      # of the digits computed so far. Returns the digit; the caller stores it back into
      # ctx before the next step. Handles zero_when / roll / reroll / adjust / clamp.
      def uwp_step(name, ctx)
        spec = (@data['uwp'] || {})[name]
        raise "ruleset #{@name.inspect} has no UWP step #{name.inspect}" unless spec
        return 0 if spec['zero_when'] && expr("uwp/#{name}/zero", spec['zero_when']).call(ctx)
        val = expr("uwp/#{name}/roll", spec['roll']).call(ctx)
        if (rr = spec['reroll']) && expr("uwp/#{name}/rr?", rr['when']).call(ctx.merge(name => val))
          val = expr("uwp/#{name}/rr=", rr['with']).call(ctx.merge(name => val))
        end
        (spec['adjust'] || []).each_with_index do |a, i|
          next unless expr("uwp/#{name}/adj#{i}", a['when']).call(ctx.merge(name => val))
          val = a.key?('set') ? a['set'].to_i : val + a['delta'].to_i
        end
        if (cl = spec['clamp'])
          val = cl.first if val < cl.first
          val = cl.last  if val > cl.last
        end
        val
      end

      private

      # Compile-and-cache an Expr by a stable key, so each formula is parsed once.
      def expr(key, src)
        (@compiled ||= {})[key] ||= Expr.compile(src)
      end
    end
  end
end

File: go-version/pkg/rules/ruleset.go
package rules

import (
	"embed"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"gopkg.in/yaml.v3"
)

//go:embed builtin/*.yml
var builtinFS embed.FS

var wordRe = regexp.MustCompile(`^\w+$`)

// Ruleset is the data-driven definition of a generation system (Traveller 5,
// Cepheus, house rules…), loaded from rules/<name>.yml. It mirrors the Ruby
// Astromapper::Rules::Ruleset: trade codes, UWP step formulas, starport/base/tech
// tables, and the algorithmic-module wiring.
type Ruleset struct {
	name       string
	data       map[string]any
	tradeOrder []string        // output order, preserved from the YAML
	tradeCond  map[string]Node // compiled trade-code conditions
	compiled   map[string]Node // cache of UWP step exprs (read-only after Load)
}

// Load resolves a ruleset by name, searching the project's rules/ dir first, then
// the embedded built-ins, and applies `extends:` inheritance before validating.
func Load(name, projectRoot string) (*Ruleset, error) {
	data, order, definesTrade, err := loadMerged(name, projectRoot, map[string]bool{})
	if err != nil {
		return nil, err
	}
	rs, err := newRuleset(name, data, order, definesTrade)
	if err != nil {
		return nil, err
	}
	if err := rs.Validate(); err != nil {
		return nil, err
	}
	return rs, nil
}

// loadMerged reads one ruleset file and folds in its parent (extends), returning the
// merged data, the resolved trade-code order, and whether trade order was set here.
func loadMerged(name, projectRoot string, seen map[string]bool) (map[string]any, []string, bool, error) {
	if seen[name] {
		return nil, nil, false, fmt.Errorf("ruleset %q has a cyclic extends chain", name)
	}
	seen[name] = true

	raw, err := readRuleFile(name, projectRoot)
	if err != nil {
		return nil, nil, false, err
	}

	var data map[string]any
	if err := yaml.Unmarshal(raw, &data); err != nil {
		return nil, nil, false, fmt.Errorf("ruleset %q: %w", name, err)
	}
	if data == nil {
		data = map[string]any{}
	}

	// Ordered trade-code keys come from whichever file declares them (`trade_codes`
	// or the wholesale-replace `trade_codes!`).
	order := orderedKeys(raw, "trade_codes!")
	definesTrade := order != nil
	if order == nil {
		order = orderedKeys(raw, "trade_codes")
		definesTrade = order != nil
	}

	if base, ok := data["extends"].(string); ok && base != "" {
		pData, pOrder, _, err := loadMerged(base, projectRoot, seen)
		if err != nil {
			return nil, nil, false, err
		}
		data = deepMerge(pData, data)
		if !definesTrade { // inherit the parent's order if the child didn't set its own
			order = pOrder
		}
	}
	return data, order, definesTrade, nil
}

// readRuleFile prefers a project override, then the embedded built-in.
func readRuleFile(name, projectRoot string) ([]byte, error) {
	if projectRoot != "" {
		p := filepath.Join(projectRoot, "rules", name+".yml")
		if b, err := os.ReadFile(p); err == nil {
			return b, nil
		}
	}
	if b, err := builtinFS.ReadFile("builtin/" + name + ".yml"); err == nil {
		return b, nil
	}
	return nil, fmt.Errorf("unknown ruleset %q (no project rules/%s.yml and no built-in)", name, name)
}

// deepMerge folds child b over parent a. A child key ending in "!" replaces the
// parent value wholesale instead of deep-merging.
func deepMerge(a, b map[string]any) map[string]any {
	out := make(map[string]any, len(a))
	for k, v := range a {
		out[k] = v
	}
	for k, bv := range b {
		if strings.HasSuffix(k, "!") {
			out[strings.TrimSuffix(k, "!")] = bv
			continue
		}
		if av, ok := out[k]; ok {
			if am, aok := av.(map[string]any); aok {
				if bm, bok := bv.(map[string]any); bok {
					out[k] = deepMerge(am, bm)
					continue
				}
			}
		}
		out[k] = bv
	}
	return out
}

// orderedKeys returns the child keys of the named top-level mapping, in document
// order (Go maps lose order, so trade-code output order is taken from the YAML node).
func orderedKeys(raw []byte, mapKey string) []string {
	var doc yaml.Node
	if yaml.Unmarshal(raw, &doc) != nil || len(doc.Content) == 0 {
		return nil
	}
	root := doc.Content[0] // mapping node
	for i := 0; i+1 < len(root.Content); i += 2 {
		if root.Content[i].Value == mapKey {
			m := root.Content[i+1]
			keys := make([]string, 0, len(m.Content)/2)
			for j := 0; j+1 < len(m.Content); j += 2 {
				keys = append(keys, m.Content[j].Value)
			}
			return keys
		}
	}
	return nil
}

func newRuleset(name string, data map[string]any, order []string, _ bool) (*Ruleset, error) {
	rs := &Ruleset{name: name, data: data, tradeOrder: order, tradeCond: map[string]Node{}, compiled: map[string]Node{}}

	// Compile trade-code conditions.
	if tc, ok := data["trade_codes"].(map[string]any); ok {
		for code, cond := range tc {
			node, err := Compile(toStr(cond))
			if err != nil {
				return nil, fmt.Errorf("ruleset %q: trade code %s: %w", name, code, err)
			}
			rs.tradeCond[code] = node
		}
	}
	// Precompile UWP step exprs so UWPStep is allocation-light and concurrency-safe.
	if uwp, ok := data["uwp"].(map[string]any); ok {
		for step, raw := range uwp {
			spec, ok := raw.(map[string]any)
			if !ok {
				continue
			}
			for _, key := range []string{"zero_when", "roll"} {
				if s, ok := spec[key].(string); ok {
					if err := rs.cache("uwp/"+step+"/"+key, s); err != nil {
						return nil, err
					}
				}
			}
			if rr, ok := spec["reroll"].(map[string]any); ok {
				for _, key := range []string{"when", "with"} {
					if s, ok := rr[key].(string); ok {
						if err := rs.cache("uwp/"+step+"/reroll/"+key, s); err != nil {
							return nil, err
						}
					}
				}
			}
			if adj, ok := spec["adjust"].([]any); ok {
				for i, a := range adj {
					if am, ok := a.(map[string]any); ok {
						if s, ok := am["when"].(string); ok {
							if err := rs.cache(fmt.Sprintf("uwp/%s/adjust/%d", step, i), s); err != nil {
								return nil, err
							}
						}
					}
				}
			}
		}
	}
	return rs, nil
}

func (rs *Ruleset) cache(key, src string) error {
	node, err := Compile(src)
	if err != nil {
		return fmt.Errorf("ruleset %q: %s: %w", rs.name, key, err)
	}
	rs.compiled[key] = node
	return nil
}

// Name is the file slug; Title is the human-facing name (the YAML `name:` field).
func (rs *Ruleset) Name() string { return rs.name }

func (rs *Ruleset) Title() string {
	if s, ok := rs.data["name"].(string); ok && s != "" {
		return s
	}
	return rs.name
}

// Validate fails fast on a malformed ruleset, reporting every problem at once.
func (rs *Ruleset) Validate() error {
	var errs []string
	if s, ok := rs.data["hex"].(string); !ok || s == "" {
		errs = append(errs, "missing `hex` alphabet")
	}
	uwp, _ := rs.data["uwp"].(map[string]any)
	for _, step := range []string{"size", "atmo", "hydro", "pop", "gov", "law"} {
		spec, ok := uwp[step].(map[string]any)
		if !ok {
			errs = append(errs, "uwp."+step+": missing")
			continue
		}
		if _, ok := spec["roll"].(string); !ok {
			errs = append(errs, "uwp."+step+": no `roll`")
		}
	}
	if sp, ok := rs.data["starport"].(map[string]any); !ok {
		errs = append(errs, "missing `starport.table`")
	} else if _, ok := sp["table"].([]any); !ok {
		errs = append(errs, "missing `starport.table`")
	}
	for _, slot := range []string{"extensions", "climate", "native"} {
		if _, err := rs.ModuleFor(slot); err != nil {
			errs = append(errs, err.Error())
		}
	}
	if len(errs) > 0 {
		return fmt.Errorf("ruleset %q is invalid:\n- %s", rs.name, strings.Join(errs, "\n- "))
	}
	return nil
}

// TradeCodes returns the world's trade classifications, in YAML (output) order.
func (rs *Ruleset) TradeCodes(ctx map[string]any) []string {
	out := []string{}
	for _, code := range rs.tradeOrder {
		if node, ok := rs.tradeCond[code]; ok && node(ctx, nil) == true {
			out = append(out, code)
		}
	}
	return out
}

// Starport returns the starport letter for an orientation roll (clamped to the table).
func (rs *Ruleset) Starport(roll int) string {
	sp, _ := rs.data["starport"].(map[string]any)
	arr, _ := sp["table"].([]any)
	if len(arr) == 0 {
		return "X"
	}
	if roll < 0 {
		roll = 0
	}
	if roll > len(arr)-1 {
		roll = len(arr) - 1
	}
	return toStr(arr[roll])
}

// TechDM sums the tech-level die modifiers from the port map and per-digit arrays.
func (rs *Ruleset) TechDM(ctx map[string]any) int {
	t, _ := rs.data["tech_dm"].(map[string]any)
	dm := 0
	if pm, ok := t["port"].(map[string]any); ok {
		if v, ok := pm[toStr(ctx["port"])]; ok {
			dm += toInt(v)
		}
	}
	for _, k := range []string{"size", "atmo", "hydro", "pop", "gov"} {
		if arr, ok := t[k].([]any); ok {
			if i := toInt(ctx[k]); i >= 0 && i < len(arr) {
				dm += toInt(arr[i])
			}
		}
	}
	return dm
}

// BaseThreshold returns the 2D threshold for a base at this starport (ok=false if
// the port can't host it).
func (rs *Ruleset) BaseThreshold(kind, port string) (int, bool) {
	b, _ := rs.data["bases"].(map[string]any)
	m, ok := b[kind].(map[string]any)
	if !ok {
		return 0, false
	}
	v, ok := m[port]
	if !ok {
		return 0, false
	}
	return toInt(v), true
}

// BaseMeets applies the ruleset's base comparison (T5 `<=`, Cepheus `>=`).
func (rs *Ruleset) BaseMeets(roll, threshold int) bool {
	op := "<="
	if b, ok := rs.data["bases"].(map[string]any); ok {
		if o, ok := b["op"].(string); ok && o != "" {
			op = o
		}
	}
	return compare(op, roll, threshold)
}

// ModuleFor returns the code module wired to an algorithmic slot (default "t5",
// constrained to a word so it only resolves to a known module).
func (rs *Ruleset) ModuleFor(slot string) (string, error) {
	name := "t5"
	if m, ok := rs.data["modules"].(map[string]any); ok {
		if v, ok := m[slot]; ok && v != nil && toStr(v) != "" {
			name = strings.ToLower(toStr(v))
		}
	}
	if !wordRe.MatchString(name) {
		return "", fmt.Errorf("ruleset %q: bad module name %q for %s", rs.name, name, slot)
	}
	return name, nil
}

// UWPStep evaluates one UWP step (size/atmo/hydro/pop/gov/law) against ctx — the
// digits computed so far — handling zero_when / roll / reroll / adjust / clamp.
func (rs *Ruleset) UWPStep(name string, ctx map[string]any, r Roller) (int, error) {
	uwp, _ := rs.data["uwp"].(map[string]any)
	spec, ok := uwp[name].(map[string]any)
	if !ok {
		return 0, fmt.Errorf("ruleset %q has no UWP step %q", rs.name, name)
	}
	if zw := rs.compiled["uwp/"+name+"/zero_when"]; zw != nil && truthy(zw(ctx, r)) {
		return 0, nil
	}
	val := toInt(rs.compiled["uwp/"+name+"/roll"](ctx, r))

	if _, ok := spec["reroll"].(map[string]any); ok {
		cw := withVal(ctx, name, val)
		if w := rs.compiled["uwp/"+name+"/reroll/when"]; w != nil && truthy(w(cw, r)) {
			val = toInt(rs.compiled["uwp/"+name+"/reroll/with"](cw, r))
		}
	}
	if adj, ok := spec["adjust"].([]any); ok {
		for i, a := range adj {
			am, ok := a.(map[string]any)
			if !ok {
				continue
			}
			cw := withVal(ctx, name, val)
			if w := rs.compiled[fmt.Sprintf("uwp/%s/adjust/%d", name, i)]; w != nil && truthy(w(cw, r)) {
				if set, ok := am["set"]; ok {
					val = toInt(set)
				} else {
					val += toInt(am["delta"])
				}
			}
		}
	}
	if cl, ok := spec["clamp"].([]any); ok && len(cl) == 2 {
		lo, hi := toInt(cl[0]), toInt(cl[1])
		if val < lo {
			val = lo
		}
		if val > hi {
			val = hi
		}
	}
	return val, nil
}

func withVal(ctx map[string]any, k string, v int) map[string]any {
	cw := make(map[string]any, len(ctx)+1)
	for key, val := range ctx {
		cw[key] = val
	}
	cw[k] = v
	return cw
}

func toStr(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}
